## Section 1:

### Task 2: ETL Process Implementation

**Dataset used**: https://archive.ics.uci.edu/dataset/352/online+retail

**Dataset Information:**

This is a transactional data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

**Variables**:

- `InvoiceNo` – (categorical) a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.	

- `StockCode` – (categorical) a 5-digit integral number uniquely assigned to each distinct product

- `Description` – (categorical) product name

- `Quantity` – (Integer) the quantities of each product (item) per transaction

- `InvoiceDate` – (Date) the day and time when each transaction was generated

- `UnitPrice` – (Continuous) product price per unit, in sterling

- `CustomerID` – (Categorical) a 5-digit integral number uniquely assigned to each customer

- `Country` – (Categorical) the name of the country where each customer resides

-------------------------------------------------------------------------

**Instruction:**

1. **Extract:** Write Python code to read the CSV file into a pandas DataFrame. Handle any missing values or data types (e.g., convert
InvoiceDate to datetime).

In [39]:
# importing necessary libraries
import pandas as pd


In [40]:
# Reading the CSV file into a pandas DataFrame
df = pd.read_excel("Online Retail.xlsx") 

# Displaying the data through '.head()'
print(f"Below are the first 5 observations in the dataset: \n")
print(df.head())


Below are the first 5 observations in the dataset: 

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  


In [41]:
# Displaying the number of rows and columns in 'Online Retail.xlsx'
num_rows, num_columns = df.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

print(f"\n")

# Describing the dataset through '.info()'
print(f"Summary of the data types and non-null counts in the dataset: \n")
print(df.info())

Number of rows: 541909
Number of columns: 8


Summary of the data types and non-null counts in the dataset: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB
None


- As seen from above, this dataset has 8 columns and 541908 rows.

- All the columns have the correct data types apart from `CustomerID` which is 'float64' (a numeric value) and should be converted to a string ('object').

- 

Duplicate rows

In [42]:
# Duplicate rows

# Checking for duplicate rows
duplicates = df.duplicated().sum()
print(f"The number of duplicate rows include: \n {duplicates} \n")

# Remove duplicates
df.drop_duplicates(inplace=True)

# Checking for duplicate rows after dropping
duplicates_after = df.duplicated().sum()
print(f"The number of duplicate rows after dropping include: \n {duplicates_after} \n")



The number of duplicate rows include: 
 5268 

The number of duplicate rows after dropping include: 
 0 



Missing Values


In [43]:
# Missing values

# Calculating the total number of missing values for each column
missing_values = df.isnull().sum()
print(f"The number of missing values in each column are: \n {missing_values} \n")

# Drop rows with missing values
df.dropna(subset=['Description', 'CustomerID'], inplace=True)

# Calculating the total number of missing values for each column after dropping
missing_values_after = df.isnull().sum()
print(f"The number of missing values in each column after dropping are: \n {missing_values_after} \n")



The number of missing values in each column are: 
 InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64 

The number of missing values in each column after dropping are: 
 InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64 



Data type conversion

In [44]:
# Converting  'InvoiceDate' column to datetime format
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Converting  'CustomerID' column to string
df['CustomerID'] = df['CustomerID'].astype(str)

# Rechecking if the columns have the correct data types through '.info()'
print(f"The data types for each column are: \n")
print(df.info())

The data types for each column are: 

<class 'pandas.core.frame.DataFrame'>
Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  object        
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 27.6+ MB
None


---------------------------------------------------------------------

**Instruction:**

2. **Transform:** 

- Calculate a new column: TotalSales = Quantity * UnitPrice.

- Create dimension-like extracts: Group by CustomerID to create a customer
summary (e.g., total purchases, country).

- Filter data for sales in the last year (assume current date as August 12, 2025).

- Handle outliers: Remove rows where Quantity < 0 or UnitPrice <= 0.

In [45]:
# Removing outliers by keeping rows where the 'Quantity' is greater that 0 and 'UnitPrice' is greater than 0
df = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)]


# Creating a new column for total sales
df['TotalSales'] = df['Quantity'] * df['UnitPrice']


# Data is between 01/12/2010 and 09/12/2011
# Confirm the range of the date in the data
print('InvoiceDate starts at:', df['InvoiceDate'].min(), 'up until', df['InvoiceDate'].max())

# Filter data for sales for last year's data 
cutoff_date = pd.to_datetime('2011-01-01')
# Obtaining records above the cutoff date to get sales for the last one year
df_recent = df[df['InvoiceDate'] > cutoff_date]


# Creating a customer dimension table
customer_dim = df_recent.groupby('CustomerID').agg({
    'TotalSales': 'sum',
    'Country': 'first'
}).reset_index()


# Creating a Time dimension table
# Extracting the date only from 'InvoiveDate'
df_recent['Date'] = df_recent['InvoiceDate'].dt.date
# Extracting the year from 'InvoiveDate'
df_recent['Year'] = df_recent['InvoiceDate'].dt.year
# Extracting the month from 'InvoiveDate'
df_recent['Month'] = df_recent['InvoiceDate'].dt.month
# Extracting the day from 'InvoiveDate'
df_recent['Day'] = df_recent['InvoiceDate'].dt.day

# Creating time_dim which stores the variables
time_dim = df_recent[['Date', 'Year', 'Month', 'Day']].drop_duplicates()



InvoiceDate starts at: 2010-12-01 08:26:00 up until 2011-12-09 12:50:00


C:\Users\Admin\AppData\Local\Temp\ipykernel_12588\2194750558.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_recent['Date'] = df_recent['InvoiceDate'].dt.date
C:\Users\Admin\AppData\Local\Temp\ipykernel_12588\2194750558.py:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_recent['Year'] = df_recent['InvoiceDate'].dt.year
C:\Users\Admin\AppData\Local\Temp\ipykernel_12588\2194750558.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row

---------------------------------------------------------------------------

**Instruction:**

3. **Load:** Use sqlite3 in Python to create a database file (retail_dw.db). Load the transformed data into a fact table (SalesFact) and at least two dimension tables (e.g.,CustomerDim, TimeDim).

In [46]:
import sqlite3

# Connect to SQLite (it will create the file if it doesn't exist)
conn = sqlite3.connect('retail_dw.db')

# Load tables
df_recent.to_sql('SalesFact', conn, if_exists='replace', index=False)
customer_dim.to_sql('CustomerDim', conn, if_exists='replace', index=False)
time_dim.to_sql('TimeDim', conn, if_exists='replace', index=False)

conn.commit()
conn.close()

print('Fact Table(SalesFact) and two dimension tables (CustomerDim and TimDim) loaded.')


Fact Table(SalesFact) and two dimension tables (CustomerDim and TimDim) loaded.


In [47]:
# Exporting the loaded tables to CSV files

conn = sqlite3.connect('retail_dw.db')

# Exporting the fact table (SalesFact)
SalesFact_df = pd.read_sql_query('SELECT * FROM SalesFact', conn)
SalesFact_df.to_csv('Data/SalesFact.csv', index=False)

# Export dimension table (CustomerDim)
CustomerDim_df = pd.read_sql_query('SELECT * FROM CustomerDim', conn)
CustomerDim_df.to_csv('Data/CustomerDim.csv', index=False)

# Export dimension table (TimeDim)
TimeDim_df = pd.read_sql_query('SELECT * FROM TimeDim', conn)
TimeDim_df.to_csv('Data/TimeDim.csv', index=False)

conn.close()
print('CSV files generated for loaded data.')

CSV files generated for loaded data.


-----------------------------------------------------------------------

**Instruction:**

4. Write a function to perform the full ETL and log the number of rows processed at each stage.

In [48]:
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)

def run_etl():
    logging.info("Starting ETL process...")

    try:
        # Extract
        df = pd.read_excel("Online Retail.xlsx")
        df.drop_duplicates(inplace=True)
        df.dropna(subset=['Description', 'CustomerID'], inplace=True)
        df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
        df['CustomerID'] = df['CustomerID'].astype(str)
        logging.info(f"Extracted {len(df)} rows.")


        # Transform
        df = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)]
        df['TotalSales'] = df['Quantity'] * df['UnitPrice']
        df_recent = df[df['InvoiceDate'] > pd.to_datetime('2011-01-01')]
        logging.info(f"{len(df_recent)} rows after filtering for last year's sales.")

        customer_dim = df_recent.groupby('CustomerID').agg({
            'TotalSales': 'sum',
            'Country': 'first'
        }).reset_index()
        logging.info(f"{len(customer_dim)} number of customer records in dimension.")

        df_recent['Date'] = df_recent['InvoiceDate'].dt.date
        df_recent['Year'] = df_recent['InvoiceDate'].dt.year
        df_recent['Month'] = df_recent['InvoiceDate'].dt.month
        df_recent['Day'] = df_recent['InvoiceDate'].dt.day
        time_dim = df_recent[['Date', 'Year', 'Month', 'Day']].drop_duplicates()
        logging.info(f"{len(time_dim)} time records in dimension.")

        # Load
        conn = sqlite3.connect('retail_dw.db')
        df_recent.to_sql('SalesFact', conn, if_exists='replace', index=False)
        customer_dim.to_sql('CustomerDim', conn, if_exists='replace', index=False)
        time_dim.to_sql('TimeDim', conn, if_exists='replace', index=False)
        conn.commit()
        conn.close()
        
        # Exporting the loaded tables to CSV
        SalesFact_df = pd.read_sql_query('SELECT * FROM SalesFact', conn)
        SalesFact_df.to_csv('Data/SalesFact.csv', index=False)
    
        CustomerDim_df = pd.read_sql_query('SELECT * FROM CustomerDim', conn)
        CustomerDim_df.to_csv('Data/CustomerDim.csv', index=False)

        TimeDim_df = pd.read_sql_query('SELECT * FROM TimeDim', conn)
        TimeDim_df.to_csv('Data/TimeDim.csv', index=False)


        logging.info("ETL process completed successfully.")

    except Exception as e:
        logging.error(f"ETL process failed: {e}")
